htmlinputelement: Update radio group validity logic for disconnected trees

elomscansio opened the pull request:
This PR updates the validity state revalidation behavior for radio input elements, ensuring consistent logic regardless of tree connectivity.

---

**What’s changed:**  
- we now always update the validity state.
- Handled more edge cases for disconnected trees, shadow roots, and cross-tree movement.
- Ensured that radio group updates still trigger when a parent is moved into a different root.
- Updated test expectations in `radio-disconnected-group-owner.html.ini`.

---

**Why this matters:**  
The updated logic aligns better with the [HTML specification](https://html.spec.whatwg.org/multipage/input.html#radio-button-group), especially regarding radio group behavior in disconnected trees or shadow DOM.

These changes improve validity propagation in scenarios like:
- Appending inputs into a disconnected form
- Shadow roots acting as radio group containers
- Cross-tree movement of radio buttons

---

**Tests:**  
- Modified `radio-disconnected-group-owner.html.ini`
  - Removed a test that now passes
  - Added a failing test expectation caused by a spec bug: the fragment parsing algorithm uses the connected tree for HTML set via `innerHTML`, affecting radio group computation

---

**Spec reference:**  
https://html.spec.whatwg.org/multipage/input.html#radio-button-group

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report an errors
- [ ] These changes fix #

<!-- Either: -->
- [X] There are tests for these changes

elomscansio commented:
Here’s a refined version of your comment with improved clarity, structure, and grammar, while preserving the technical detail and your line of thought:

---

Hi @jdm, I'm currently working on fixing the failing test in  
`tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html`.

I’m a bit confused by what seems like a contradiction in the test behavior.  
At [[line 180](https://github.com/servo/servo/blob/2fe57cc2a2e5e7b982f7a32349117e2516724682/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L180)](https://github.com/servo/servo/blob/2fe57cc2a2e5e7b982f7a32349117e2516724682/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L180), the test expects the radio button to remain checked, but I initially assumed that the later insertion should override the previous check.

The test suite comments suggest that revalidation **shouldn’t** happen when the tree is disconnected. However, this appears to conflict with line 202: https://github.com/servo/servo/blob/2fe57cc2a2e5e7b982f7a32349117e2516724682/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L202, which seems to expect a change in `valueMissing`.

Is there a meaningful distinction between a "disconnected tree" and a "disconnected form" in this context?

For reference, I added a check to skip revalidation for radio buttons whose root is disconnected from the tree:  
see here: https://github.com/elomscansio/servo/blob/af5cb191121c2e905c3b1220121134a64d5bf783/components/script/dom/htmlinputelement.rs#L2583-L2588

Would appreciate any clarification!

elomscansio commented on components/script/dom/htmlinputelement.rs:
```diff
@@ -1257,6 +1257,7 @@ impl HTMLInputElementMethods<crate::DomTypeHolder> for HTMLInputElement {
     // https://html.spec.whatwg.org/multipage/#dom-input-checked
     fn SetChecked(&self, checked: bool) {
         self.update_checked_state(checked, true);
+        update_related_validity_states(self, CanGc::note())
```
when `radio.checked = true|false` i.d direct assignment `SetChecked` is being executed from codeGen script binding, so i thought it would be best if we update the validity state here. this actually made one of the failing test passed

jdm replied:
Good point!

elomscansio commented:
Hello @jdm still waiting for your response to know if these discrepancy is the right behaviour, so i can continue debugging.

elomscansio commented:
> > Is there a meaningful distinction between a "disconnected tree" and a "disconnected form" in this context?
> 
> Yes, a `<form>` element is a [form owner](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#form-owner) for the radio elements, whereas the `<div>` is not a form owner. From the [specification](https://html.spec.whatwg.org/multipage/input.html#radio-button-state-(type=radio)):
> 
> ```
> When any of the following phenomena occur, if the element's checkedness state is true after the occurrence, the checkedness state of all the other elements in the same radio button group must be set to false:
> ```
> 
> In that list is `The element's form owner changes.` which is why the two tests expect different behaviour.

Thanks for the clearification

elomscansio commented:
tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html

I have been able to get other test passed but not 
1. [line 48 - 70](https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L48-L70): I added check for disconnected tree (components/script/dom/htmlinputelement.rs:1744-1787) but the test expect revalidation in a disconnected tree. i'm kinda confuse on how to go about it
2. [line 166 - 186](https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L166-L186): will always fail as:
binding to tree execute three times 
E1. from [components/script/dom/element.rs:2160](https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/components/script/dom/element.rs#L2160) and it uses connected root to parse htmlString
E2. from [components/script/dom/element.rs:2170-2177](https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/components/script/dom/element.rs#L2170-L2177) by appending to document fragment
E3. from [components/script/dom/element.rs:3045](https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/components/script/dom/element.rs#L3045) by replace operation

first our test in bind_to_tree for disconnected tree will fail because E1 uses a connected tree, even though the real bind_to_tree is E3 replace operation.

----

And some direct property assignment will not trigger revalidation for example `elem.required = true|false` won't trigger revalidation see
1. [components/script/dom/macros.rs:260-270](https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/components/script/dom/macros.rs#L260-L270) macro
2. [components/script/dom/htmlinputelement.rs:1541](https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/components/script/dom/htmlinputelement.rs#L1541)
3. [tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html:54](https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L54)

elomscansio commented:
Should i submit the current changes or is there a way to resolve the remaining test issue

jdm commented:
>[line 48 - 70](https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L48-L70): I added check for disconnected tree ([components/script/dom/htmlinputelement.rs:1744-1787](https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/components/script/dom/htmlinputelement.rs#L1744-L1787)) but the test expect revalidation in a disconnected tree. i'm kinda confuse on how to go about it

This big set of checks looks like it's trying to special case a bunch of test scenarios, while I believe the actual logic we need to implement is [this](https://html.spec.whatwg.org/multipage/input.html#radio-button-group):
* always attempt to revalidate
* to iterate over the radio group, start from the root element of the given element and iterate over all reachable children
* the radio group is all radio input elements that have the same form owner (possibly null)

jdm commented:
I've been staring at https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L166-L186 for a while and I still can't figure out why E1 in servo behaves differently than other browsers (who all [pass the test](https://wpt.fyi/results/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html?label=master&label=experimental&aligned&q=radio-disconnected-group-owner.html)). I think we should file an issue about it to keep track of it and ignore the failure for now.

jdm commented:
Oh, looks like ladybird is having similar difficulties: https://github.com/whatwg/html/issues/11023

elomscansio commented:
> I've been staring at https://github.com/elomscansio/servo/blob/7b118f1da693bb8c9729449edaa93e154816b57f/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L166-L186 for a while and I still can't figure out why E1 in servo behaves differently than other browsers (who all [pass the test](https://wpt.fyi/results/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html?label=master&label=experimental&aligned&q=radio-disconnected-group-owner.html)). I think we should file an issue about it to keep track of it and ignore the failure for now.

ok, how do we file the issue

elomscansio commented:
> Oh, looks like ladybird is having similar difficulties: [whatwg/html#11023](https://github.com/whatwg/html/issues/11023) . Well done @elomscansio on identifying your first specification bug :)

Thanks

elomscansio commented on components/script/dom/htmlinputelement.rs:
```diff
@@ -2552,7 +2611,12 @@ impl VirtualMethods for HTMLInputElement {
         }
 
-        update_related_validity_states(self, can_gc);
+        if !pushing_attribute {
```
hello @jdm could you please take a look at the skipping of group validation when pushing attribute, to know if I should remove the skipping

jdm replied:
Why do we want to skip it? Also, can't this trigger when setting an attribute while the element is already connected to the tree?

elomscansio replied:
WHY: I thought that we are doing unnecessary group revalidation, that radio_group_iter will return all the radio inputs, but i realize that if the element is not yet connected it just return it self so there no need of skipping .

And looking at https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/components/script/dom/element.rs#L2613 might resolve to `AttributeMutation::Set(None)`

So skipping it is wrong, i will remove it right away

jdm commented on components/script/dom/htmlinputelement.rs:
```diff
@@ -2579,8 +2643,15 @@ impl VirtualMethods for HTMLInputElement {
             .check_ancestors_disabled_state_for_form_control();
-
-        update_related_validity_states(self, can_gc);
+        if context.is_in_tree() || should_update_validity_state(self) {
+            if self.input_type() == InputType::Radio {
+                self.radio_group_updated(self.radio_group_name().as_ref());
+            }
+            update_related_validity_states(self, can_gc);
+        } else {
+            self.validity_state()
+                .perform_validation_and_update(ValidationFlags::all(), can_gc);
```
I'm concerned here that this logic is more complicated than necessary. should_update_validity_state is too specific, and it feels like it's compensating for some other code that is wrong. I don't know why we're passing `None` in [this code](https://github.com/servo/servo/blob/24e329838c8f01dedc5e92b42332078928282ddb/components/script/dom/htmlinputelement.rs#L1702) but it looks extremely suspicious to me. Try passing `Some(root)` there, and then the code highlighted here could probably become:
```rust
        if self.input_type() == InputType::Radio {
            self.radio_group_updated(self.radio_group_name().as_ref());
        }
        update_related_validity_states(self, can_gc);
```

elomscansio replied:
`should_update_validity_state` is used to handle special cases where group revalidation is still required. While [[line 186](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L186)](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L186) in the test suggests we **should not** perform revalidation when the radio group is in a disconnected tree, several earlier lines — [[70](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L70)](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L70), [[93](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L93)](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L93), [[122](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L122)](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L122), [[144](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L144)](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L144), and [[164](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L164)](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L164) — show that validity should still be updated in specific edge cases. This function tries to detect those exceptions to ensure correctness.

elomscansio replied:
OR does the test different from spec i'm not sure i really understand the spec and the test

4.10.5.1.16 Radio Button state (type=radio)
✔MDN
When an [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) element's [type](https://html.spec.whatwg.org/multipage/input.html#attr-input-type) attribute is in the [Radio Button](https://html.spec.whatwg.org/multipage/input.html#radio-button-state-(type=radio)) state, the rules in this section apply.

The [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) element [represents](https://html.spec.whatwg.org/multipage/dom.html#represents) a control that, when used in conjunction with other [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) elements, forms a [radio button group](https://html.spec.whatwg.org/multipage/input.html#radio-button-group) in which only one control can have its [checkedness](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-checked) state set to true. If the element's [checkedness](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-checked) state is true, the control represents the selected control in the group, and if it is false, it indicates a control in the group that is not selected.

The radio button group that contains an [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) element a also contains all the other [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) elements b that fulfill all of the following conditions:

The [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) element b's [type](https://html.spec.whatwg.org/multipage/input.html#attr-input-type) attribute is in the [Radio Button](https://html.spec.whatwg.org/multipage/input.html#radio-button-state-(type=radio)) state.
Either a and b have the same [form owner](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#form-owner), or they both have no [form owner](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#form-owner).
Both a and b are in the same [tree](https://dom.spec.whatwg.org/#concept-tree).
They both have a [name](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-name) attribute, their [name](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-name) attributes are not empty, and the value of a's [name](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-name) attribute equals the value of b's [name](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-name) attribute.
A [tree](https://dom.spec.whatwg.org/#concept-tree) must not contain an [input](https://html.spec.whatwg.org/multipage/input.html#the-input-element) element whose [radio button group](https://html.spec.whatwg.org/multipage/input.html#radio-button-group) contains only that element.

When any of the following phenomena occur, if the element's [checkedness](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-checked) state is true after the occurrence, the [checkedness](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-checked) state of all the other elements in the same [radio button group](https://html.spec.whatwg.org/multipage/input.html#radio-button-group) must be set to false:

The element's [checkedness](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#concept-fe-checked) state is set to true (for whatever reason).
The element's [name](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fe-name) attribute is set, changed, or removed.
The element's [form owner](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#form-owner) changes.
[A type change is signalled](https://html.spec.whatwg.org/multipage/input.html#signal-a-type-change) for the element.
The element [becomes connected](https://html.spec.whatwg.org/multipage/infrastructure.html#becomes-connected).

elomscansio commented:
Just a ping

jdm commented:
Yeah, sorry for the delay. This is the kind of review that I need to sit down for several hours and focus on what the tests are doing and what our implementation is doing, and I haven't been able to make that happen yet.

jdm commented:
Let's focus on [this subtest](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L20-L46) right now:
* the first failure on the main branch happens at [this valueMissing check](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L41), which you fixed by updating the validity status in SetChecked
* the second failure happens at [this `checked` check](https://github.com/servo/servo/blob/d724c8e9e3b4a9ff945bb82202a2ae53ed440fbe/tests/wpt/tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html#L44), which you fixed by calling `self.radio_group_updated` in `bind_to_tree`

If I apply those minimal changes in the following diff against the main branch:
```diff
diff --git a/components/script/dom/htmlinputelement.rs b/components/script/dom/htmlinputelement.rs
index 1999c7193f..8f9819718b 100644
--- a/components/script/dom/htmlinputelement.rs
+++ b/components/script/dom/htmlinputelement.rs
@@ -1257,6 +1257,7 @@ impl HTMLInputElementMethods<crate::DomTypeHolder> for HTMLInputElement {
     // https://html.spec.whatwg.org/multipage/#dom-input-checked
     fn SetChecked(&self, checked: bool) {
         self.update_checked_state(checked, true);
+        update_related_validity_states(self, CanGc::note());
     }

     // https://html.spec.whatwg.org/multipage/#dom-input-readonly
@@ -2580,6 +2581,9 @@ impl VirtualMethods for HTMLInputElement {
         self.upcast::<Element>()
             .check_ancestors_disabled_state_for_form_control();

+        if self.input_type() == InputType::Radio {
+            self.radio_group_updated(self.radio_group_name().as_ref());
+        }
         update_related_validity_states(self, can_gc);
     }
```

I get:
```
UNEXPECTED-PASS Removed elements are moved into separate radio groups.
FAIL Appending input radio input into a disconnect tree don't update the other radio inputs in the same radio group. - UNEXPECTED-PASS Appending input radio input into a disconnect form should update the other radio inputs in the same radio group.
```

Given the specification bug that you previously identified (https://github.com/whatwg/html/issues/11023), I am not surprised that the `Appending input radio input into a disconnect tree don't update the other radio inputs in the same radio group.` test now fails. I propose that we reduce this PR to the diff that I pasted above (which is based on your original changes), then mark the two tests as passing in the INI file and the other test as failing.

jdm commented:
These changes also cause:
```
  ▶ OK /html/semantics/forms/the-input-element/radio.html
  └   ▶ PASS [expected FAIL] Radio buttons in an orphan tree should make a group
```
which is great!

elomscansio committed:
htmlinputelement: Update radio group validity logic for disconnected trees

The validity state update logic for radio input elements is now performed
unconditionally, aligning more closely with the HTML spec, so that validity
recalculations occur consistently across tree contexts.

This improves support for validity propagation when:
- Inputs are inside disconnected documents or shadow roots
- Inputs are moved across trees
- Radio buttons are appended into disconnected forms

Additionally, we handle a potential edge case where a parent node is moved
into a different root, ensuring group updates are still triggered.

We also updated the WPT .ini file:
- Removed the passing expectation for a previously failing test
- Added an expectation for a failing test caused by a spec bug where the
  fragment parsing algorithm uses the connected tree to parse HTML strings
  set via `innerHTML`, affecting radio group containment logic

Spec reference:
https://html.spec.whatwg.org/multipage/input.html#radio-button-group

Signed-off-by: Emmanuel Paul Elom <elomemmanuel007@gmail.com>

elomscansio commented:
Requested changes applied

jdm reviewed (approved):
Thanks for digging into these test failures! I didn't realize it would turn into such a complicated investigation!
